### PR TITLE
Migrate to edition 2021

### PR DESCRIPTION
**Pull request by makr11st**
## Proposed changes

As the latest rust moved to `edition 2021` here is bump of all current crates to use edition 2021.
All tests seem to be working fine.

All code seems to be ok with the change.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Review by didier-wenzek — COMMENTED** [review]
I'm happy with this upgrade. Let's merge it just after the 0.5 release.

File: crates/common/batcher/Cargo.toml [hunk 1]
```diff
@@ -2,7 +2,8 @@
 name = "batcher"
 version = "0.5.1"
 authors = ["thin-edge.io team <info@thin-edge.io>"]
-edition = "2018"
+edition = "2021"
+rust-version = "1.58"
```
Comment (didier-wenzek) [inline crates/common/batcher/Cargo.toml:6]: Do we really need `1.58` as minimum `rust-version` ?

* For now we use none of these new features - from yesterday ;-).
* That said, I agree that using "Disjoint capture in closures" and "Captured identifiers in format strings" will be really helpful.

File: .github/workflows/commit-workflow.yml [hunk 2]
```diff
@@ -208,4 +208,4 @@ jobs:
         # https://github.com/marketplace/actions/rust-cargo
         with:
           command: msrv
-          args: --minimum 1.54.0
+          args: --minimum 1.58.0
```
Comment (rina23q) [inline .github/workflows/commit-workflow.yml:211]: I think we can get rid of `Cargo msrv --version` step completely.

File: crates/core/thin_edge_json/benches/parsing.rs [hunk 3]
```diff
@@ -41,7 +41,7 @@ impl MeasurementVisitor for DummyVisitor {
 }
 
 fn parse_stream(input: &str) {
-    thin_edge_json::stream::parse_str(input, &mut DummyVisitor).unwrap();
```
Comment (rina23q) [inline crates/core/thin_edge_json/benches/parsing.rs:44]: Is it related to this PR?

File: crates/core/c8y_translator/fuzz/Cargo.toml [hunk 4]
```diff
@@ -4,7 +4,7 @@ name = "c8y_translator-fuzz"
 version = "0.0.0"
 authors = ["Automatically generated"]
 publish = false
-edition = "2018"
+edition = "2021"
```
Comment (rina23q) [inline crates/core/c8y_translator/fuzz/Cargo.toml:8]: We miss `rust-version`.
```suggestion
rust-version = "1.58"
```

File: .github/workflows/hosted-post-merge_version_update.yml [hunk 5]
```diff
@@ -20,7 +20,7 @@ jobs:
       - name: enable toolchain via github action
         uses: actions-rs/toolchain@v1
         with:
-          toolchain: 1.54
+          toolchain: 1.58.0
```
Comment (rina23q) [inline .github/workflows/hosted-post-merge_version_update.yml:23]: Here specifies one more digit than before. I guess the reason, in our commit workflow we specified 3 digit like `1.54.0`, and keep here and the commit workflow consistent?

**Review by rina23q — COMMENTED** [review]
Left some comments.

**Comment by rina23q** [comment]
`cargo clippy` fails, but it makes sense.
I think you need to divide this PR into two. 1) workflow changes only 2) Cargo.toml change.
If we merge 1) first, we can avoid this kind of errors.

```
/home/runner/.cargo/bin/cargo clippy
error: package `json-writer v0.5.1 (/home/runner/work/thin-edge.io/thin-edge.io/crates/common/json_writer)` cannot be built because it requires rustc 1.58 or newer, while the currently active rustc version is 1.57.0
Error: The process '/home/runner/.cargo/bin/cargo' failed with exit code 101
```

**Review by didier-wenzek — APPROVED** [review]
Approved

**Pull request merged**